### ISSUE TITLE
False positive reportPossiblyUnboundVariable with pattern matching

### ISSUE THREAD

**Issue by tamird** [microsoft/pyright#8214]
```python
async def MyFunctionThatReturnsSequence() -> Sequence[Any]:
    return []


async def MyFunctionThatTakesFoo(foo: Any) -> None:
    pass


async def bar() -> None:
    foos = await MyFunctionThatReturnsSequence()
    assert isinstance(foos, list)  # added just in case this was the reason for pyright complaint.

    match foos:
        case []:
            raise RuntimeError("bad")
        case [foo]:
            pass
        case [foo, *_]:
            print("Multiple Foos found")

    await MyFunctionThatTakesFoo(foo)  # "foo" is possibly unbound
```

Seen both on the CLI (pyright 1.1.368) and in pylance (v2024.6.1).

**Comment by erictraut** [comment]
Could you please provide a minimal, self-contained example? The code above references symbols that are not defined.

**Comment by tamird** [comment]
Updated.

**Comment by erictraut** [comment]
Thanks for the updated code sample.

Pyright is working as intended in this case, so I don't consider this a bug.

Pyright (and other static type checkers) use type narrowing for pattern matching statements. In your example, the initial type of `obj` is `Sequence[Any]`. This cannot be narrowed any further by any of the cases, so this is considered a non-exhaustive match from a static type checking perspective. 

This is an unfortunate limitation of the Python static type system, not something that is specific to pyright. Other type checkers (like mypy) would also consider this a non-exhaustive match. The number of elements in a `Sequence` are not included in the type that describes the sequence, so there's no way to indicate a type such as "a sequence that is at least one element in length".

Here's a way you can update the code to work around this limitation:

```python
    match foos:
        case [foo]:
            pass
        case [foo, *_]:
            print("Multiple Foos found")
        case _:
            raise RuntimeError("bad")
```
